blocks/pipeline.py

```diff
@@ -9,6 +9,31 @@
 from utils.flatten import flatten
 
 from .base import Block, DataSource, Element
+from type import DataType
+
+from utils.process_block import process_block
+
+
+class Sequence(Block):
```

Comment:
is this a block that we execute? or only something we return from Hierarchy?
and... what does a Sequence do?

Reply:
There is a conceptual question/confusion for me on the fact that we treated ``Concat`` as ``DataSource``. 
So I did this:

```python
full_pipeline = Pipeline(
    "nlp_hf_meta-model-pipeline",
    VectorConcat(
        "concat-source", [huggingface_distilroberta, huggingface_baseline_distilbert]
    ),
    remove_none(
        [
            ListOfListsToNumpy(replace_nan=True),
            SKLearnModel("sklearn-meta-model", sklearn_config),
        ]
    ),
)
```

Instead of the more logical (but not because Concat doesn't inherit from ``Block`` and is not model, it is not operational):

```python
concat_pipeline = Pipeline(
    "concat-pipeline",
    input_data,
    [
        VectorConcat(
            "concat-source",
            [huggingface_distilroberta, huggingface_baseline_distilbert],
        )
    ],
)

full_pipeline = Pipeline(
    "nlp_hf_meta-model-pipeline",
    concat_pipeline,
    remove_none(
        [
            ListOfListsToNumpy(replace_nan=True),
            SKLearnModel("sklearn-meta-model", sklearn_config),
        ]
    ),
)
```

**Back to your question:**
So this ``Sequence`` is only used for getting the hierarchy and appending the right name in the hierarchy. 
Because we treat concat as a datasource we cannot properly wrap it in a pipeline -> we need to create an object that wraps the otherwise implicitly defined wrapper.

Reply:
In the second example it is odd that the concat needs an datasource:
```python

concat_pipeline = Pipeline(
    "concat-pipeline",
    input_data,
    [
        VectorConcat(
            "concat-source",
            [huggingface_distilroberta, huggingface_baseline_distilbert],
        )
    ],
)```

Whereas the ``datasources`` are already defined under the individual pipelines that ``concat`` receives.

Reply:
alright!
so, then my next question is:
why do we need a Sequential class?
Why can't we just return the Pipeline as self, and 
`                children=[source_hierarchy, current_pipeline_hierarchy],` as the children?

I think there's a reason why but I don't yet understand it, that's why I'm asking!

Reply:
Then we define the pipeline we are in as both the parent and the child:
```
 - pipeline we are invoking from
           - datasource objects
           - pipeline we are invoking from
 ```
We could treat it as the child of ``datasource`` but that would be misleading because its on the same level
```
- datasource
        - pipeline we are invoking from
```
or treat it as the child of the pipeline we are invoking from (maybe this is equally valid!).
```
- pipeline we are invoking from
       - datasource (previous pipelines)
```

Reply:
The last example is actually wrong, because it messes up the sequence:
```
- pipeline we are invoking from
       - datasource (previous pipelines)
```
we need this:
```
- Pipeline
       - Pipeline we are in invoking from
       - Datasource (previous pipelines)
```

``Sequence`` is basically pipelines without the sophisticated machinery that it has (eg.: you have to give it models)

Reply:
@itchingpixels